src/Classes/Canvas.ts

```diff
@@ -154,7 +158,159 @@ export class Canvas {
             this.paint();
         } else {
             if (this.context) {
-                this.context.clearRect(0, 0, this.dimension.width, this.dimension.height);
+                CanvasUtils.clear(this.context, this.dimension);
+            }
+        }
+    }
+
+    public drawPolygonMask(rawData: ICoordinates[]): void {
+        const container = this.container;
+        const options = container.options;
+        const context = this.context;
+        const polygonDraw = options.polygon.draw;
+
+        if (context) {
+            CanvasUtils.drawPolygonMask(context, rawData, polygonDraw.lineColor, polygonDraw.lineWidth);
+        }
+    }
+
+    public drawLinkedLine(p1: Particle, p2: Particle, pos1: ICoordinates, pos2: ICoordinates, opacity: number): void {
+        const container = this.container;
+        const options = container.options;
+
+        const ctx = this.context;
+
+        if (!ctx) {
+            return;
+        }
+
+        let colorLine: IRgb | undefined;
+
+        /*
+         * particles connecting line color:
+         *
+         *  random: in blink mode : in every frame refresh the color would change
+         *          hence resulting blinking of lines
+         *  mid: in consent mode: sample particles color and get a mid level color
+         *                        from those two for the connecting line color
+         */
+
+        if (container.particles.lineLinkedColor === "random") {
+            colorLine = Utils.getRandomColorRGBA();
+        } else if (container.particles.lineLinkedColor == "mid" && p1.color && p2.color) {
+            const sourceColor = p1.color;
+            const destColor = p2.color;
+
+            colorLine = {
+                b: Math.floor(Utils.mixComponents(sourceColor.b, destColor.b, p1.radius, p2.radius)),
+                g: Math.floor(Utils.mixComponents(sourceColor.g, destColor.g, p1.radius, p2.radius)),
+                r: Math.floor(Utils.mixComponents(sourceColor.r, destColor.r, p1.radius, p2.radius)),
+            };
+        } else {
+            colorLine = container.particles.lineLinkedColor as IRgb;
+        }
+
+        const width = container.retina.lineLinkedWidth;
+
+        CanvasUtils.drawLineLinked(ctx, width, pos1, pos2, options.backgroundMask.enable, colorLine, opacity);
+    }
+
+    public drawConnectLine(p1: Particle, p2: Particle): void {
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Declaration of public instance method not allowed after declaration of private instance method. Instead, this should come after constructors.](https://app.codacy.com/manual/ar3s/tsparticles/pullRequest?prid=5171185)